xds/internal/balancer/cdsbalancer/cdsbalancer_test.go

```diff
@@ -315,507 +322,1272 @@ func setupWithWatch(t *testing.T) (*fakeclient.Client, *cdsBalancer, *testEDSBal
+	select {
+	case err := <-resolverErrCh:
+		if err != resolverErr {
+			t.Fatalf("Error pushed to child policy is %v, want %v", err, resolverErr)
+		}
+	case <-ctx.Done():
+		t.Fatal("Timeout when waiting for resolver error to be pushed to the child policy")
 	}
-	if err := counter.StartRequest(maxRequests); err == nil {
-		t.Fatal("unexpected success on start request over max")
+
+	// Ensure that the ClientConn moves to TransientFailure.
+	for state := cc.GetState(); state != connectivity.TransientFailure; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatalf("Timed out waiting for state change. got %v; want %v", state, connectivity.TransientFailure)
+		}
 	}
-	counter.EndRequest()
 }
 
-// TestClose verifies the Close() method in the CDS balancer.
+// Tests that closing the cds LB policy results in the cluster resource watch
+// being cancelled and the child policy being closed.
 func (s) TestClose(t *testing.T) {
-	grpctest.TLogger.ExpectError("cds-lb.*Received balancer config after close")
+	// Register a wrapped cds LB policy for the duration of this test that
+	// allows explicitly closing the LB policy.
+	cdsBuilder := balancer.Get(cdsName)
+	internal.BalancerUnregister(cdsBuilder.Name())
+	cdsBalancerCh := make(chan balancer.Balancer, 1)
+	stub.Register(cdsBuilder.Name(), stub.BalancerFuncs{
+		Init: func(bd *stub.BalancerData) {
+			bal := cdsBuilder.Build(bd.ClientConn, bd.BuildOptions)
+			bd.Data = bal
+			cdsBalancerCh <- bal
+		},
+		ParseConfig: func(lbCfg json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+			return cdsBuilder.(balancer.ConfigParser).ParseConfig(lbCfg)
+		},
+		UpdateClientConnState: func(bd *stub.BalancerData, ccs balancer.ClientConnState) error {
+			bal := bd.Data.(balancer.Balancer)
+			return bal.UpdateClientConnState(ccs)
+		},
+		Close: func(bd *stub.BalancerData) {
+			bal := bd.Data.(balancer.Balancer)
+			bal.Close()
+		},
+	})
+	t.Cleanup(func() { balancer.Register(cdsBuilder) })
+
+	// Register a wrapped clusterresolver LB policy (child policy of the cds LB
+	// policy) for the duration of this test that closes a channel when closed.
+	clusterresolverBuilder := balancer.Get(clusterresolver.Name)
+	internal.BalancerUnregister(clusterresolverBuilder.Name())
+	childPolicyCloseCh := make(chan struct{})
+	stub.Register(clusterresolver.Name, stub.BalancerFuncs{
+		Init: func(bd *stub.BalancerData) {
+			bd.Data = clusterresolverBuilder.Build(bd.ClientConn, bd.BuildOptions)
+		},
+		ParseConfig: func(lbCfg json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+			return clusterresolverBuilder.(balancer.ConfigParser).ParseConfig(lbCfg)
+		},
+		UpdateClientConnState: func(bd *stub.BalancerData, ccs balancer.ClientConnState) error {
+			bal := bd.Data.(balancer.Balancer)
+			return bal.UpdateClientConnState(ccs)
+		},
+		Close: func(bd *stub.BalancerData) {
+			bal := bd.Data.(balancer.Balancer)
+			bal.Close()
+			close(childPolicyCloseCh)
+		},
+	})
```

Comment:
Same comment here. Perhaps a universal helper that returns three channels corresponding to UpdateCCS, ResolverError, and Close and just use the channel you need for each test that uses helper.

Reply:
Done. Done. Done.